### PR TITLE
nixos/hadoop: add hadoop module (hdfs, yarn)

### PR DESCRIPTION
**Pull request by aespinosa**
###### Motivation for this change

Create initial Hadoop modules: HDFS and YARN.  Enough to make a small working Hadoop cluster.

The modules are currently minimal that it doesn't include other HDFS, YARN or MapReduce components to make it "production-ish"

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by aespinosa** [comment]
just realized that the NixOS tests only works in Hadoop 3.0 onwards. I originally cherry-picked this PR from #40712 (which is Hadoop 3.1.0)

Hadoop 3.0.0 changed the default ports in HDFS and YARN.

**Comment by volth** [comment]
HDFS service (and tests) can also fail with pre-3.0.0 hadoop because of hardcoded paths like "/bin/ls" in Java code

That's why I build Hadoop from sources after fixing those spots (https://github.com/NixOS/nixpkgs/pull/32893).

it is fixed only in 3.0.0 and binary distribution is ok since then

**Comment by volth** [comment]
Maybe ```hadoop``` and ```hadoop-bin``` are to be made akin to ```firefox```/```firefox-bin```

**Comment by aespinosa** [comment]
@GrahamcOfBorg test hadoop.hdfs

**Comment by aespinosa** [comment]
@GrahamcOfBorg test hadoop.yarn

**Comment by matthewbauer** [comment]
@GrahamcOfBorg test hadoop.yarn hadoop.hdfs

**Comment by matthewbauer** [comment]
@GrahamcOfBorg test hadoop.yarn

**Comment by matthewbauer** [comment]
@GrahamcOfBorg test hadoop.hdfs

**Comment by matthewbauer** [comment]
Looks like there is an odd eval error. Does the test run locally>

**Comment by aespinosa** [comment]
let me check on the fresh rebase. it was building locally before

**Comment by aespinosa** [comment]
@matthewbauer it should evaluate now

**Comment by matthewbauer** [comment]
@GrahamcOfBorg test hadoop.yarn hadoop.hdfs

**Comment by aespinosa** [comment]
@matthewbauer I think i fixed the last missing part of my rebase. tests should pass now.

File: nixos/modules/services/cluster/hadoop/default.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      example = literalExample "pkgs.hadoop";
+      description = ''
+      '';
+    };
+  };
+
+
+  config = mkMerge [
+    (mkIf (builtins.hasAttr "yarn" config.users.extraUsers ||
+           builtins.hasAttr "hdfs" config.users.extraUsers ) {
```
Comment (aszlig) [inline nixos/modules/services/cluster/hadoop/default.nix]: @aespinosa: How come this is still using `extraUsers`/`extraGroups`? Do we have some outdated documentation somewhere we need to fix?

**Pull request merged**